lib/ansible/modules/web_infrastructure/ansible_tower/tower_host.py

```diff
@@ -82,46 +96,74 @@ def main():
     )
     module = TowerModule(argument_spec=argument_spec, supports_check_mode=True)
 
     name = module.params.get('name')
-    description = module.params.get('description')
     inventory = module.params.get('inventory')
-    enabled = module.params.get('enabled')
-    state = module.params.get('state')
+    state = module.params.pop('state')
+    merge_variables = module.params.pop('merge_variables')
+    groups = module.params.pop('groups')
 
     variables = module.params.get('variables')
-    if variables:
-        if variables.startswith('@'):
-            filename = os.path.expanduser(variables[1:])
-            with open(filename, 'r') as f:
-                variables = f.read()
 
     json_output = {'host': name, 'state': state}
 
     tower_auth = tower_auth_config(module)
     with settings.runtime_values(**tower_auth):
         tower_check_mode(module)
         host = tower_cli.get_resource('host')
+        group_res = tower_cli.get_resource('group')
+        params = module.params.copy()
+        params['create_on_missing'] = True
 
         try:
             inv_res = tower_cli.get_resource('inventory')
             inv = inv_res.get(name=inventory)
+            params['inventory'] = inv['id']
+        except exc.notFound as excinfo:
+            module.fail_json(msg='Failed to update the host, inventory not found: {0}'.format(excinfo), changed=False)
+
+        group_ids = []
+        try:
+            for group in groups:
+                group_ids.append(group_res.get(name=group, inventory=params['inventory'])['id'])
+        except exc.NotFound as excinfo:
+            module.fail_json(msg='Group {0} does not exist: {1}'.format(group, excinfo), changed=False)
 
+        try:
+            existing_host = host.get(name=name, inventory=params['inventory'])
+        except exc.NotFound:
+            existing_host = None
+            json_output['created'] = True
+            module.log("Existing host not found, will create")
+
+        try:
+            if merge_variables and existing_host:
+                params['variables'] = sanitise_and_merge_variables(existing_host['variables'], variables)
+            else:
+                params['variables'] = sanitise_and_merge_variables(variables)
+        except TypeError as excinfo:
+            module.fail_json(msg='Invalid variable data {0}'.format(excinfo))
+
+        try:
             if state == 'present':
-                result = host.modify(name=name, inventory=inv['id'], enabled=enabled,
-                                     variables=variables, description=description, create_on_missing=True)
+                result = host.modify(**params)
                 json_output['id'] = result['id']
+                json_output['changed'] = result['changed']
+                for group in group_ids:
+                    res = host.associate(host=result['id'], group=group)
+                    if res['changed']:
+                        json_output['changed'] = True
```

Comment:
Looking at https://github.com/ansible/ansible/pull/47971/files, which attempts the same thing, perhaps it would be good to include `groups_added` in the output, so give data on what new groups have been associated. However, I wonder if a list of all groups may be more helpful.